File: packages/lcyt-backend/src/routes/sync.js
import { Router } from 'express';

/**
 * Factory for the /sync router.
 *
 * POST /sync — Trigger an NTP-style clock sync for the session's sender.
 * Requires JWT auth (passed as middleware).
 *
 * Accepts both session JWTs (with sessionId) and project-scoped JWTs (with apiKey).
 *
 * @param {import('../store.js').SessionStore} store
 * @param {import('express').RequestHandler} auth - Pre-created auth middleware
 * @returns {Router}
 */
export function createSyncRouter(store, auth) {
  const router = Router();

  // POST /sync — Clock synchronization (auth required)
  router.post('/', auth, async (req, res) => {
    const { sessionId, apiKey, projectId } = req.session;

    // Resolve session: prefer sessionId if available (session JWT),
    // otherwise try apiKey or projectId (project-scoped JWT)
    const resolvedSessionId = sessionId || apiKey || projectId;
    let session = null;

    if (sessionId) {
      // Direct session lookup using sessionId from session JWT
      session = store.get(sessionId);
    } else if (apiKey || projectId) {
      // Project-scoped JWT: find a session for this project/apiKey
      // Search through all sessions to find one matching the apiKey
      const projectKey = apiKey || projectId;
      for (const [sid, sess] of store.entries()) {
        if (sess.apiKey === projectKey) {
          session = sess;
          break;
        }
      }
    }

    if (!session) {
      return res.status(404).json({ error: 'Session not found' });
    }

    try {
      const syncResult = await session.sender.sync();

      // Update session syncOffset with the newly computed value
      session.syncOffset = syncResult.syncOffset;
      store.touch(sessionId);

      return res.status(200).json({
        syncOffset: syncResult.syncOffset,
        roundTripTime: syncResult.roundTripTime,
        serverTimestamp: syncResult.serverTimestamp,
        statusCode: syncResult.statusCode
      });
    } catch (err) {
      return res.status(502).json({
        error: err.message || 'Sync failed: YouTube server did not respond',
        statusCode: 502
      });
    }
  });

  return router;
}

File: python-packages/lcyt-backend/lcyt_backend/routes/sync.py
"""POST /sync — NTP-style clock sync for the session's sender (Bearer auth)."""

import time

from flask import Blueprint, current_app, g, jsonify

from ..middleware.auth import require_auth

sync_bp = Blueprint("sync", __name__)


@sync_bp.post("/")
@require_auth
def sync_session():
    """POST /sync — Trigger a clock sync for the current session."""
    senders = current_app.config["SENDERS"]
    session_id = g.session["sessionId"]
    entry = senders.get(session_id)

    if not entry:
        return jsonify({"error": "Session not found"}), 404

    try:
        sender = entry["sender"]
        t0 = time.monotonic()
        result = sender.heartbeat()
        t1 = time.monotonic()

        rtt_ms = int((t1 - t0) * 1000)
        sync_offset = rtt_ms // 2

        entry["sync_offset"] = sync_offset

        return jsonify({
            "syncOffset": sync_offset,
            "roundTripTime": rtt_ms,
            "serverTimestamp": result.server_timestamp,
            "statusCode": result.status_code,
        }), 200

    except Exception as exc:
        return jsonify({
            "error": str(exc) or "Sync failed: YouTube server did not respond",
            "statusCode": 502,
        }), 502
